tianshou/data/batch.py

```diff
@@ -502,49 +546,92 @@ def to_torch(self, dtype: Optional[torch.dtype] = None,
                 v.to_torch(dtype, device)
 
     def cat_(self,
-             batches: Union['Batch', List[Union[dict, 'Batch']]]) -> None:
+             batches: Union['Batch', List[Union[dict, 'Batch']]],
+             lens: Optional[List[int]] = None
+             ) -> None:
         """Concatenate a list of (or one) :class:`~tianshou.data.Batch` objects
         into current batch.
+        If x=Batch(a=Batch(a=np.random.randn(3, 4)), b=np.random.randn(3, 4)),
+        y = Batch(a=Batch(a=Batch()), b=np.random.randn(3, 4)), if we want to
+        concatenate x and y, we want to pad y.a.a with zeros. Without ``lens``
+        as a hint, when we concatenate x.a and y.a, we would not be able to
+        know how to pad y.a. So we must have the argument ``lens`` to give us
+        a hint.
         """
         if isinstance(batches, Batch):
             batches = [batches]
         if len(batches) == 0:
             return
         batches = [x if isinstance(x, Batch) else Batch(x) for x in batches]
-        if len(self.__dict__) > 0:
+        if lens is None:
+            # remove empty Batch, and infer lens
+            batches = [x for x in batches if not x.is_empty()]
```

Comment:
Why not to use only the `try...catch` black ?

Reply:
Can you elaborate on that? I do not get it.

Reply:
I was just wonderting if this line is necessary since you are using recursive `is_empty` in the try/catch block.

Reply:
Oh, this is necessary. ``is_empty()`` and ``is_empty(recursive=True)`` mean different things and have to be dealt with differently.

For ``is_empty() == True``, this is ``Batch()``, and will be ignored because this is a top-level cat.

For ``is_empty(recursive=True) == True``, this is nested empty Batch but not Batch(), and we have to treat it as length zero and keep it.

Reply:
OK nice.

Reply:
Maybe add a small comment at the end of those lines to make it clear.